pandas/core/reshape/merge.py

```diff
@@ -2158,8 +2158,18 @@ def _factorize_keys(
         # GH#23917 TODO: needs tests for case where lk is integer-dtype
         #  and rk is datetime-dtype
         klass = libhashtable.Int64Factorizer
-        lk = ensure_int64(np.asarray(lk))
-        rk = ensure_int64(np.asarray(rk))
+
+        # both types are equal to each other
+        if lk.dtype != rk.dtype:
+            # lk is not a null type, so the left value should be analyzed
+            if lk.dtype != pd.Int64Dtype():
```

Comment:
We can not do this like that. We also have Int64Dtype for example. This has to be more general.